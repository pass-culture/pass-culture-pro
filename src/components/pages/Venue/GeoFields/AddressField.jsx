/* eslint
  react/jsx-one-expression-per-line: 0 */
import classnames from 'classnames'
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import { composeValidators } from 'react-final-form-utils'

import Address from './Address'
import FieldErrors from 'components/layout/form/FieldErrors'
import validateRequiredField from 'components/layout/form/utils/validateRequiredField'

const createBatchGeoChanges = form => values => {
  form.batch(() => {
    const {
      address,
      city,
      latitude,
      longitude,
      postalCode,
      selectedAddress,
    } = values
    form.change('address', address)
    form.change('city', city)
    form.change('latitude', latitude)
    form.change('longitude', longitude)
    form.change('postalCode', postalCode)
    form.change('selectedAddress', selectedAddress)
  })
}

export const AddressField = ({
  className,
  disabled,
  form,
  format,
  id,
  innerClassName,
  label,
  name,
  parse,
  placeholder,
  readOnly,
  renderInner,
  renderValue,
  required,
  validate,
  ...AddressProps
}) => {
  const requiredValidate =
    required && typeof required === 'function'
      ? required
      : (required && validateRequiredField) || undefined

  const batchGeoChanges = createBatchGeoChanges(form)

  return (
    <Field
      format={format}
      name={name}
      validate={composeValidators(validate, requiredValidate)}
      render={({ input, meta }) => {
        return (
          <div
            className={classnames('field text-field', className, {
              'is-label-aligned': label,
              'is-read-only': readOnly,
            })}
            id={id}>
            <label
              htmlFor={name}
              className={classnames('field-label', { empty: !label })}>
              {label && (
                <span>
                  <span>{label}</span>
                  {required && !readOnly && (
                    <span className="field-asterisk">*</span>
                  )}
                </span>
              )}
            </label>
            <div className="field-control">
              <div className="field-value flex-columns items-center">
                <div
                  className={classnames(
                    'field-inner flex-columns items-center',
                    innerClassName
                  )}>
                  <Address
                    {...input}
                    {...AddressProps}
                    className="field-input field-address"
                    disabled={disabled || readOnly}
                    name={name}
                    onMarkerDragend={batchGeoChanges}
                    onSuggestionSelect={batchGeoChanges}
                    onTextChange={batchGeoChanges}
                    placeholder={readOnly ? '' : placeholder}
                    readOnly={readOnly}
                    required={!!required}
                  />
                  {renderInner()}
                </div>
                {renderValue()}
              </div>
              <FieldErrors meta={meta} />
            </div>
            <div />
          </div>
        )
      }}
    />
  )
}

AddressField.defaultProps = {
  className: '',
  disabled: false,
  format: null,
  id: null,
  innerClassName: null,
  label: '',
  parse: null,
  placeholder: '',
  readOnly: false,
  renderInner: () => null,
  renderValue: () => null,
  required: false,
  type: 'text',
  validate: null,
}

AddressField.propTypes = {
  autoComplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  form: PropTypes.object.isRequired,
  format: PropTypes.func,
  id: PropTypes.string,
  innerClassName: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  parse: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderInner: PropTypes.func,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  validate: PropTypes.func,
}

export default AddressField
