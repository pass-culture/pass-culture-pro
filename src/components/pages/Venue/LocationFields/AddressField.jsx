/* eslint
  react/jsx-one-expression-per-line: 0 */
import classnames from 'classnames'
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import { composeValidators } from 'react-final-form-utils'

import LocationViewer from './LocationViewer'
import FieldErrors from 'components/layout/form/FieldErrors'
import getRequiredValidate from 'components/layout/form/utils/getRequiredValidate'

const createBatchLocationChanges = (form, isLocationFrozen) => location => {
  form.batch(() => {
    const { address, city, latitude, longitude, postalCode } = location
    form.change('address', address)
    form.change('city', city)
    form.change('isLocationFrozen', isLocationFrozen)
    form.change('latitude', latitude)
    form.change('longitude', longitude)
    form.change('postalCode', postalCode)
  })
}

export const AddressField = ({
  className,
  disabled,
  form,
  format,
  id,
  innerClassName,
  label,
  name,
  parse,
  placeholder,
  readOnly,
  renderInner,
  renderValue,
  required,
  validate,
  ...AddressProps
}) => (
  <Field
    format={format}
    name={name}
    validate={composeValidators(validate, getRequiredValidate(required))}
    render={({ input, meta }) => {
      return (
        <div
          className={classnames('field text-field', className, {
            'is-label-aligned': label,
            'is-read-only': readOnly,
          })}
          id={id}>
          <label
            htmlFor={name}
            className={classnames('field-label', { empty: !label })}>
            {label && (
              <span>
                <span>{label}</span>
                {required && !readOnly && (
                  <span className="field-asterisk">*</span>
                )}
              </span>
            )}
          </label>
          <div className="field-control">
            <div className="field-value flex-columns items-center">
              <div
                className={classnames(
                  'field-inner flex-columns items-center',
                  innerClassName
                )}>
                <LocationViewer
                  {...input}
                  {...AddressProps}
                  className="field-input field-address"
                  disabled={disabled || readOnly}
                  name={name}
                  onMarkerDragend={createBatchLocationChanges(form, false)}
                  onSuggestionSelect={createBatchLocationChanges(form, true)}
                  onTextChange={createBatchLocationChanges(form, false)}
                  placeholder={readOnly ? '' : placeholder}
                  readOnly={readOnly}
                  required={!!required}
                />
              </div>
            </div>
            <FieldErrors meta={meta} />
          </div>
          <div />
        </div>
      )
    }}
  />
)

AddressField.defaultProps = {
  className: '',
  disabled: false,
  format: null,
  id: null,
  innerClassName: null,
  label: '',
  parse: null,
  placeholder: '',
  readOnly: false,
  required: false,
  type: 'text',
  validate: null,
}

AddressField.propTypes = {
  autoComplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  form: PropTypes.object.isRequired,
  format: PropTypes.func,
  id: PropTypes.string,
  innerClassName: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  parse: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderInner: PropTypes.func,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  validate: PropTypes.func,
}

export default AddressField
