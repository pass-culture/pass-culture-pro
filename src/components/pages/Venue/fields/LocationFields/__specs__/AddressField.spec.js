import AddressField from '../AddressField'
import React from 'react'
import { shallow } from 'enzyme'

describe('src | components | pages | Venue | fields | AddressField', () => {
  let props

  beforeEach(() => {
    props = {
      name: 'fake name',
      form: {}
    }
  })

  describe('snapshot', () => {
    it('should match snapshot', () => {
      // when
      const wrapper = shallow(<AddressField {...props} />)

      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
