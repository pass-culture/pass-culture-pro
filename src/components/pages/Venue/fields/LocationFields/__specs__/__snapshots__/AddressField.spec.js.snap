// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Venue | fields | AddressField addressFieldRender should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    AddressProps={Object {}}
    className="fake className"
    disabled={false}
    form={Object {}}
    id={1}
    innerClassName="fake inner className"
    label="fake label"
    name="fake name"
    placeholder="fake placeholder"
    readOnly={true}
    required={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): [Function],
  Symbol(enzyme.__nodes__): Array [
    [Function],
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`src | components | pages | Venue | fields | AddressField snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddressField
    className=""
    disabled={false}
    form={Object {}}
    format={null}
    id={null}
    innerClassName={null}
    label=""
    name="fake name"
    placeholder=""
    readOnly={false}
    required={false}
    validate={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "format": null,
      "name": "fake name",
      "render": [Function],
      "validate": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "format": null,
        "name": "fake name",
        "render": [Function],
        "validate": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
